### ISSUE TITLE
Cross-compiling from Fedora to Windows using MinGW64

### ISSUE THREAD

**Issue by allquixotic** [rust-lang/rust#67587]
On Fedora and Red Hat-derived systems (CentOS, etc.) there is a cross-compilation toolchain with the compiler triplet `x86_64-w64-mingw32`. Here are a few facts about it:

 - It has been supported at least since RHEL 7 / CentOS 7 and for about a dozen releases of Fedora
 - The packages are regularly updated
 - Fedora ships many free and open source software middleware packages as installable binaries via `yum` / `dnf`; installing these packages puts the appropriate headers, link libraries, pkg-config files, etc. into the right place within the "sys-root" mingw64, which lives in `/usr/x86_64-w64-mingw32/sys-root`.
 - Fedora offers several "wrapper" commands that drive well-known build systems. They have `mingw64-configure`, which automatically applies arguments and environment variables so GNU autotools can compile seamlessly with this toolchain. They have `mingw64-cmake`, for cmake. And they have `mingw64-env`, which just exports about 30 environment variables to try to make a "generic" build system aware of this toolchain. It exports `PKG_CONFIG_PATH`, `CC`, `CPP`, `CXX`, `DLLTOOL`, `GCC`, `WINDRES`, `STRIP`, `XML2_CONFIG`, `CXXFILT`, `CXX`, `LD`, `RANLIB`, and so on. For many build systems other than rust, this is enough to get them compiling native code even without explicit support in the build system for the compiler triplet.
 - The built binaries produce native code for 64-bit Windows operating systems, but there's also a toolchain available for 32-bit Windows with a different compiler triplet (starts with i686).

I've found this toolchain very convenient for native code in the past, because most build systems can just treat this toolchain "as if it were Linux" (with some caveats, like possibly needing to run `windres` and other Windows-only tools) and 95% of C/C++ code that gives even a basic consideration to cross-platform support will just compile and work, especially if it depends on deliberately cross-platform and Windows-friendly lower level libraries for the heavy lifting.

On Fedora 31 using rustc 1.40 for the native Linux system from `rustup` , `rustc --print target-list` doesn't list this compiler triplet. Can support be added for it without too much work? I think rustc already supports **a** MinGW toolchain; it's just that the current MinGW toolchain support expects the system compiling the code to be Windows as well. In my case, the system doing the compiling would be a modern 64-bit Linux distro and the target system of the built binaries would be 64-bit Windows.

**Comment by crlf0710** [comment]
Usually the major problem is that their exception models are different. MinGW comes with different exception "flavors" and the one flavor that normal "rust-pc-windows-gnu" picks is different from what Linux distros ships. If they're consistent i think you can just use the built-in target to cross compile modulo a few linker path settings etc.

**Comment by retep998** [comment]
Remember to do `rustup target add x86_64-pc-windows-gnu` and then specify `--target x86_64-pc-windows-gnu` in your `cargo` commands.

**Comment by mati865** [comment]
Fedora 31 still uses SJLJ for i686 but they are switching to DWARF-2 in the next Fedora release. That means only x86_64 will work for now.
Mingw target is called `x86_64-pc-windows-gnu` (LLVM convention) and after doing what retep998 said and configuring the linker in cargo config you should have working cross compilation for Windows.

**Comment by allquixotic** [comment]
I was able to cross-compile to Windows successfully using the MinGW64 toolchain in Fedora 31 (for 64-bit; I have no use for 32-bit binaries at all) using the steps retep998 suggested.

It's much appreciated that this "just works" with a little bit of setup. Is it worth keeping this issue open to discuss improving documentation? If not, it can be closed, as my immediate problem is solved. For the record, I was able to compile `librespot` (https://github.com/librespot-org/librespot) for 64-bit Windows from Fedora 31's MinGW64 toolchain. I was also able to make source modifications to add a Gstreamer backend for it, which also compiles and runs correctly. It's all very nice.
